bridges/otellogr/logsink.go

```diff
@@ -0,0 +1,405 @@
+func (l *LogSink) Info(level int, msg string, keysAndValues ...any) {
+	l.log(nil, msg, l.levelSeverity(level), keysAndValues...)
+}
+
+// Init receives optional information about the logr library this
+// implementation does not use it.
+func (l *LogSink) Init(info logr.RuntimeInfo) {
+	// We don't need to do anything here.
+	// CallDepth is used to calculate the caller's PC.
+	// PC is dropped as part of the conversion to the OpenTelemetry log.Record.
+}
+
+// WithName returns a new LogSink with the specified name appended.
+func (l LogSink) WithName(name string) logr.LogSink {
+	l.logger = l.newLogger(name)
+	return &l
+}
+
+// WithValues returns a new LogSink with additional key/value pairs.
+func (l LogSink) WithValues(keysAndValues ...any) logr.LogSink {
+	_, attrs := convertKVs(keysAndValues)
+	l.values = append(l.values, attrs...)
+	return &l
+}
+
+// convertKVs converts a list of key-value pairs to a list of [log.KeyValue].
+// The last [context.Context] value is returned as the context.
+func convertKVs(keysAndValues []any) (context.Context, []log.KeyValue) {
+	ctx := context.Background()
+
+	if len(keysAndValues) == 0 {
+		return ctx, nil
+	}
+	if len(keysAndValues)%2 != 0 {
+		// Ensure an odd number of items here does not corrupt the list
+		keysAndValues = append(keysAndValues, nil)
+	}
+
+	kv := make([]log.KeyValue, 0, len(keysAndValues)/2)
+	for i := 0; i < len(keysAndValues); i += 2 {
+		k, ok := keysAndValues[i].(string)
+		if !ok {
+			// Ensure that the key is a string
+			k = fmt.Sprintf("%v", keysAndValues[i])
+		}
+
+		v := keysAndValues[i+1]
+		if vCtx, ok := v.(context.Context); ok {
+			// Special case when a field is of context.Context type.
+			ctx = vCtx
+			continue
+		}
+
+		kv = append(kv, log.KeyValue{
+			Key:   k,
+			Value: convertValue(v),
+		})
+	}
+	return ctx, kv
+}
+
+func convertValue(v any) log.Value {
+	// Handling the most common types without reflect is a small perf win.
+	switch val := v.(type) {
+	case bool:
+		return log.BoolValue(val)
+	case string:
+		return log.StringValue(val)
+	case int:
+		return log.Int64Value(int64(val))
+	case int8:
+		return log.Int64Value(int64(val))
+	case int16:
+		return log.Int64Value(int64(val))
+	case int32:
+		return log.Int64Value(int64(val))
+	case int64:
+		return log.Int64Value(val)
+	case uint:
+		return convertUintValue(uint64(val))
+	case uint8:
+		return log.Int64Value(int64(val))
+	case uint16:
+		return log.Int64Value(int64(val))
+	case uint32:
+		return log.Int64Value(int64(val))
+	case uint64:
+		return convertUintValue(val)
+	case uintptr:
+		return convertUintValue(uint64(val))
+	case float32:
+		return log.Float64Value(float64(val))
+	case float64:
+		return log.Float64Value(val)
+	case time.Duration:
+		return log.Int64Value(val.Nanoseconds())
+	case complex64:
+		return log.StringValue(strconv.FormatComplex(complex128(val), 'f', -1, 64))
+	case complex128:
+		return log.StringValue(strconv.FormatComplex(val, 'f', -1, 128))
```

Comment:
Can we translate it in the same way as `otelzap`?
Sorry, but I have not noticed this before.

Reply:
For reference: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/cb45b36c145ff13e4ba76fd9e02ef39297c906b5/bridges/otelzap/encoder.go#L90-L94